src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.cc

```diff
@@ -0,0 +1,604 @@
+        std::make_unique<Extent>(
+          punch_pt->get_key(),
+          val.laddr,
+          val.lextent_offset,
+          val.length));
+      journal_remove(punch_pt, maybe_get_delta_buffer());
+      deflist.push_back(val.laddr);
+      continue;
+    }
+    // deref head
+    logger().debug("ExtMapLeafNode::punch_head {} {}", lo, len);
+    uint32_t keep = punch_pt->get_key() + val.length - loend;
+    ret.emplace_back(
+      std::make_unique<Extent>(
+        punch_pt->get_key(),
+        val.laddr,
+        val.lextent_offset,
+        val.length - keep));
+    lext_map_val_t inval(val.laddr, val.lextent_offset + val.length -keep, keep);
+    journal_insert(punch_pt + 1, loend, inval, maybe_get_delta_buffer());
+    journal_remove(punch_pt, maybe_get_delta_buffer());
+    break;
+  }
+  return crimson::do_for_each(deflist.begin(), deflist.end(), [this, &t, &ret]
+	 (const auto &laddr) {
+    return tm->dec_ref(t, laddr).safe_then([](auto i) {
+      return punch_lextent_ertr::now();
+    });
+  }).safe_then ([this, ret = std::move(result_up)] {
+    return punch_lextent_ertr::make_ready_future<extent_map_list_t>(std::move(*ret));
+  });
+}
+
+ExtMapLeafNode::rm_lextent_ret
+ExtMapLeafNode::rm_lextent(Transaction &t, objaddr_t lo, lext_map_val_t val)
+{
+  if(!is_pending()) {
+    auto mut =  make_duplicate(t, this)->cast<ExtMapLeafNode>();
+    return mut->rm_lextent(t, lo, val);
+  }
+
+  auto [rm_pt, rm_end] = get_leaf_entries(lo, val.length);
+  if(lo == rm_pt->get_key() && val.laddr == rm_pt->get_val().laddr && val.length == rm_pt->get_val().length){
+    journal_remove(rm_pt, maybe_get_delta_buffer());
+    logger().debug(
+      "ExtMapLeafNode::rm_lextent: removed {}->{} {} {}",
+      rm_pt.get_key(),
+      rm_pt.get_val().laddr,
+      rm_pt.get_val().lextent_offset,
+      rm_pt.get_val().length);
+      return rm_lextent_ertr::make_ready_future<bool>(true);
+  } else {
+    return rm_lextent_ertr::make_ready_future<bool>(false);
+  }
+}
+
+ExtMapLeafNode::find_hole_ret
+ExtMapLeafNode::find_hole(Transaction &t, objaddr_t lo, extent_len_t len)
+{
+  logger().debug("ExtMapLeafNode::find_hole lo={}, len={}, *this={}",
+                 lo, len, *this);
+  auto [start, end] = bound(lo, lo + len);
+
+  //cross at least one mapping
+  if (start != end && start++ != end)
+    return find_hole_ertr::make_ready_future<ExtentRef>(
+		    std::make_unique<Extent>(lo, L_ADDR_NULL, 0, 0));
+  //overlap at start tail
+  if (start != end && lo < (start->get_key() + start->get_val().length))
+    return find_hole_ertr::make_ready_future<ExtentRef>(
+                    std::make_unique<Extent>(lo, L_ADDR_NULL, 0, 0));
+  //overlap at start head
+  if (start != end && start->get_key() >= lo)
+    return find_hole_ertr::make_ready_future<ExtentRef>(
+                    std::make_unique<Extent>(lo, L_ADDR_NULL, 0, 0));
+
+  if (start != begin())
+    start = start - 1;
+
+  return tm->read_extents<TestBlock>(t, start->get_val().laddr,
```

Comment:
I don't think I understand what find_hole is meant to do?

Reply:
To expand on this: it may be that adding a more detailed explanation could solve it, but there are some things that aren't obvious to me:
- I don't see a case where the interface user would need to find an unmapped extent (hole)
- I don't understand ExtMapLeafNode::find_hole:
  - Why is it valid to read the extent at at start->get_val() as a TestBlock (that extent generally won't be a TestBlock)?
  - Why would this component need to read the mapped block?
  - I don't really understand the L_ADDR_NULL returns early in the method